### ISSUE TITLE
Emotion / CSS bug

### ISSUE THREAD

**Issue by Robdel12** [kark/percy-snapshot#1]
Hey @kark! This doesn't appear to be a Percy issue but more of Percy catching a bug. Looking at this it appears there's an issue with the inline styles that are being applied here: 

https://github.com/kark/percy-snapshot/blob/master/src/App.tsx#L26

When it renders it renders like this:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/2072894/189956374-348d09e5-215f-427d-9fee-8f9083d0208d.png">

There's missing values for the border radius inline styles. On top of that, there's zero rules in the CSSOM's rulesheet:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/2072894/189957022-d198b28e-0437-48c9-83b0-4262837d0ced.png">

I'm not sure how these styles are being applied. There's no rules in the CSSOM, the inline styles are clearly invalid, and there's no styles in the linked app stylesheet either. Given all of that, once the DOM snapshot is captured (where CSSOM would be serialized, but it doesn't have any rules in it), Percy renders the page "correctly" with the broken inline CSS rules. 

If the output into the DOM was this, everything would work:

<img width="361" alt="image" src="https://user-images.githubusercontent.com/2072894/189958091-59d4a5bf-6a80-4a6f-bd84-6350ba683ea0.png">

So to sum up, there's some kind of bug going on here. I spent about an hour and a half trying to figure out how those styles are being applied but I can't find it. But from Percy's perspective, it's rendering "correctly" (since that inline style in the DOM is invalid and there's no CSSOM rules to capture).

**Comment by kark** [comment]
Hi [Robdel12](https://github.com/Robdel12)!

Thank you for creating this thread.

Yes, I agree that there is a specific browser behaviour happening [here](https://github.com/kark/percy-snapshot/blob/master/src/App.tsx#L26) as you mentioned.  

Personally I wouldn’t say Percy catches a bug here, I would rather call it “not replicating real browser(s) behavior”.

> But from Percy's perspective, it's rendering "correctly" (since that inline style in the DOM is invalid and there's no CSSOM rules to capture).

Please mind that from the Percy user’s perspective this is not rendering correctly (since the output in the browser is different).

Even though the values are empty, the computed values exist and are applied. 
<img width="644" alt="Screenshot 2022-09-14 at 09 27 11" src="https://user-images.githubusercontent.com/49066275/190094386-68ad0906-a9e4-4855-8e01-1fa3c2984218.png">
<img width="1098" alt="Screenshot 2022-09-14 at 09 29 20" src="https://user-images.githubusercontent.com/49066275/190094407-a0a4d5a2-0c48-4667-902d-a5aa6d439b3b.png">

Please note that the following code fragment:
```html
<h1
  style={{
    backgroundColor: "var(--color)",
    borderRadius: "var(--br)",
    height: "calc(2*50px)",
    borderBottomRightRadius: 0,
  }}
>
  Hello Percy 👋
</h1>
```
does not contain any invalid html nor css. What browsers yield (the expanded version of border-radius) is not something we can control, but we expect Percy to display it the way the browser does (and update your rerender flow of the serialized DOM accordingly).

**Comment by Robdel12** [comment]
Hey @kark! 

> Personally I wouldn’t say Percy catches a bug here, I would rather call it “not replicating real browser(s) behavior”.

I don't necessarily agree. 

> Even though the values are empty, the computed values exist and are applied.

This is exactly what the bug I'm talking about is. It's quite peculiar that both the CSSOM is _empty_ and the inline styles in the DOM are _invalid_. The code fragment does not contain invalid HTML or CSS, that's correct. But what is outputted and rendered into the DOM (which is what Percy captures) is _not_ valid. In order for this to work, we'll need either valid CSS in the DOM (as the inline styles) or [styles in the CSSOM for us to serialize.](https://docs.percy.io/docs/debugging-sdks#cssom-rules) Right now, neither of those are true so Percy can't capture the right state. The bug is exactly that. There are no CSSOM rules and the inline styles in the DOM are invalid. Whatever the JS is doing to half apply these styles is a bug. 

If you don't care to address the bug and want to paper it over with enabling JavaScript, you'll have to make your application be able to encode it's state so when its re-rendered, it's not getting replaced. Right now the toggle of the theme is in-memory only, so when the JS runs over the page in Percy's browsers it resets it to the initial state (and not the toggled state).

**Comment by kark** [comment]
Hi @Robdel12!

Thanks for your reply.

>> Personally I wouldn’t say Percy catches a bug here, I would rather call it “not replicating real browser(s) behavior”.

> I don't necessarily agree.

Chrome and Firefox behave consistently. Safari doesn't even yield any invalid style rules. Visual output is exactly the same in all three. Percy shows something different.

> But what is outputted and rendered into the DOM (which is what Percy captures) is not valid.

Well, OK, but this seems like something with your flow internally (snapshot collection and rendering, something that browser does is lost in between). There is a quirk for sure, but from the black-box perspective, we get a different visual output from Percy than from the browsers.

> If you don't care to address the bug and want to paper it over with enabling JavaScript, ...

We do care. 
It's not that we want something, we asked for advice how to best use your tool and how can this issue be addressed.
We didn't come up with this solution, it was suggested by Percy support team.
We didn't find much in [Percy docs](https://docs.percy.io/docs/cli-configuration#snapshot) and we didn't get answer how this option influences the render process in details. This quirk only happens in case of JS-based styling (css vars seem to be applied in the visual tests when the values are assigned in document's css and css modules), therefore it made sense to us to use this option.

Please mind that the bug is not in our code, and we need a solution to move forward. That is why we contacted you.
We can't get rid of css vars. We can't afford not to use `border` and `border-radius`.

So the question we have now is what can be done here? Will this behaviour be addressed in any way by Percy?
If your point is that we should wait until this is addressed by Chrome and Firefox, than that's not really an option for us.

**Comment by Robdel12** [comment]
Hey @kark! I think there's a misunderstanding about where the bug actually is. It's not the browser but the CSS library that's being used here to generate the styles (Emotion). The bug doesn't have anything to do with CSS vars (those fully work, when they're actually set).

Given this:

```js
style={{
  borderRadius: "var(--br)",
  borderBottomRightRadius: 0,
}}
```

The correct output into the DOM would be:

```html
background-color: var(--color);
border-top-left-radius: var(--br); 
border-top-right-radius: var(--br); 
border-bottom-right-radius: 0px; 
border-bottom-left-radius: var(--br); 
```

But what actually happens is:

```html
background-color: var(--color);
border-top-left-radius: ; 
border-top-right-radius: ; 
border-bottom-right-radius: 0px; 
border-bottom-left-radius: ; 
```

Notice how `background-color` gets the proper `var` set, but not the `border-radius-*`. How the styles are somehow set, in memory somewhere, I don't know. But it's not in the inline styles and it's also not in the CSSOM. 

If you remove `borderBottomRightRadius: 0`, then what is output is valid and works:

<img width="1721" alt="image" src="https://user-images.githubusercontent.com/2072894/190443089-a68d82d1-ff44-41c7-8ce9-e108b98d49f9.png">

This also produces a proper screenshot in Percy with JS disabled:

<img width="1510" alt="image" src="https://user-images.githubusercontent.com/2072894/190443648-4034deea-ca3f-4432-aceb-7006f3c50e67.png">

However Emotion is handling splitting the radii out, it's missing the `var(--br)` value in the inline styles.

**Comment by kark** [comment]
Hey @Robdel12!

Hmm, yes, I guess there is a misunderstanding. In my opinion, it has nothing to do with `@emotion/react` but its root cause is the use of css vars.

I guess it might have been misleading that `@emotion/react` is in dependencies (sorry for that, I reused this repro repository and I should have removed all the redundant code). The thing is that `@emotion/react` is not used in the app. Please mind that on `master` branch only React-based styling is used.

Please checkout `vanilla-js` branch - https://github.com/kark/percy-snapshot/pull/2. The issue can be observed even with vanilla JS.

Therefore, the quirk (radii splitting out) is a pure browser thing (and how Percy displays it).
Again, sorry for the confusion.

**Comment by Robdel12** [comment]
Ah gotcha, I never checked the dependencies of the project.

> Therefore, the quirk (radii splitting out) is a pure browser thing (and how Percy displays it).

 This is a React bug then, not a pure browser thing (or more likely a combo of the two). Whatever is generating this HTML/inline style is broken. This is still true:

```
style={{
  borderRadius: "var(--br)",
  borderBottomRightRadius: 0,
}}
```

The correct output into the DOM would be:

```
background-color: var(--color);
border-top-left-radius: var(--br); 
border-top-right-radius: var(--br); 
border-bottom-right-radius: 0px; 
border-bottom-left-radius: var(--br); 
```

But what actually happens is:

```
background-color: var(--color);
border-top-left-radius: ; 
border-top-right-radius: ; 
border-bottom-right-radius: 0px; 
border-bottom-left-radius: ; 
```

This can even be replicated with `margin` & `marginRight`. See this codesandbox: https://codesandbox.io/s/intelligent-davinci-9phj2y?file=/src/App.js

And even more so, if you remove the CSS var, React stops spitting out invalid styles:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/2072894/190499071-52134019-5792-44e0-878d-8407dd5040c6.png">

**Comment by kark** [comment]
Hey,
Thanks for sharing.
Unfortunately, this is not my perspective. This is 100% pure browser thing as I see it.
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <style>
      .App {
        font-family: sans-serif;
        text-align: center;
      }
    </style>
  </head>
  <body>
    <div class="App">
      <button id="button">change theme</button>
      <h1>Hello Percy 👋</h1>
    </div>
    <script>
      const h1 = document.querySelector("h1");
      const button = document.querySelector("#button");
      const root = document.querySelector(":root");

      window.addEventListener("load", () => {
        root.style.setProperty("--br", "20px");
        root.style.setProperty("--color", "palegoldenrod");
        h1.style.backgroundColor = "var(--color)";
        h1.style.borderRadius = "var(--br)";
        h1.style.height = "calc(2*50px)";
        h1.style.borderBottomRightRadius = 0;
      });
      button.addEventListener("click", () => {
        root.style.setProperty("--br", "60px");
        root.style.setProperty("--color", "tomato");
        root.style.setAttribute("data-theme", "dark");
      });
    </script>
  </body>
</html>
```
If you'd save it as `html` file and open in:
**Chrome**
<img width="972" alt="Screenshot 2022-09-16 at 08 16 36" src="https://user-images.githubusercontent.com/49066275/190570037-19f6170f-9336-431f-bcdd-bd760ba0d422.png">
**FF**
<img width="1239" alt="Screenshot 2022-09-16 at 08 17 06" src="https://user-images.githubusercontent.com/49066275/190570104-589124e1-282e-44d7-9931-0a1787358abc.png">
**Safari (no invalid css rules here)**
<img width="1169" alt="Screenshot 2022-09-16 at 08 18 12" src="https://user-images.githubusercontent.com/49066275/190570137-8ab446c7-df6b-4435-b6fe-1753aad21ff7.png">

So from the browser perspective one might say about a quirk. But this should be addressed solely by Percy.

**Comment by Robdel12** [comment]
> But this should be addressed solely by Percy.

I don’t see how you think that’s the case here when there’s zero bugs from Percy, but bugs from react. I’m going to close this since I don’t think we’ll ever agree.
